tutorials/src/bin/basic-tutorial-9.rs

```diff
@@ -0,0 +1,150 @@
+            print!("{}: ", tag);
+            values.for_each(|v| {
+                // reviewer: this isn't nice, but if you call
+                // serialize() on a string it escapes it since it has
+                // to be able to read it back again, which makes for
+                // ugly looking output. Is there a better way?
+                if let Some(s) = v.downcast_ref::<&str>() {
+                    println!("{}", s.get().unwrap())
+                } else if let Some(serialized) = v.serialize() {
+                    println!("{}", serialized);
+                }
+            });
+        })
+    };
+}
+
+/* Print information regarding a stream and its substreams, if any */
+fn print_topology(info: &DiscovererStreamInfo, depth: usize) {
+    print_stream_info(info, depth);
+
+    if let Some(next) = info.get_next() {
+        print_topology(&next, depth + 1);
+    } else if let Some(container_info) = info.downcast_ref::<DiscovererContainerInfo>() {
+        for stream in container_info.get_streams() {
+            print_topology(&stream, depth + 1);
+        }
+    }
+}
+
+fn on_connect_discovered(
+    _discoverer: &Discoverer,
+    discoverer_info: &DiscovererInfo,
+    error: &Option<glib::Error>,
+) {
+    let uri = discoverer_info.get_uri().unwrap();
+    match discoverer_info.get_result() {
+        DiscovererResult::Ok => println!("Discovered {}", uri),
+        DiscovererResult::UriInvalid => println!("Invalid uri {}", uri),
+        DiscovererResult::Error => if let Some(msg) = error {
+            println!("{}", msg);
+        },
+        DiscovererResult::Timeout => println!("Timeout"),
+        DiscovererResult::Busy => println!("Busy"),
+        DiscovererResult::MissingPlugins => if let Some(s) = discoverer_info.get_misc() {
+            println!("{}", s);
+        },
+        _ => println!("Unknown result"),
+    }
+
+    if discoverer_info.get_result() != DiscovererResult::Ok {
+        return;
+    }
+
+    println!("Duration: {}", discoverer_info.get_duration());
+
+    if let Some(tags) = discoverer_info.get_tags() {
+        println!("Tags:");
+        tags.iter_tag_list().for_each(|(tag, values)| {
+            print!("  {}: ", tag);
+            values.for_each(|v| {
+                if let Some(s) = v.downcast_ref::<&str>() {
+                    println!("{}", s.get().unwrap())
+                } else if let Some(serialized) = v.serialize() {
+                    println!("{}", serialized);
+                }
+            });
+        })
+    };
+
+    println!(
+        "Seekable: {}",
+        if discoverer_info.get_seekable() {
+            "yes"
+        } else {
+            "no"
+        }
+    );
+
+    println!("Stream information:");
+
+    if let Some(stream_info) = discoverer_info.get_stream_info() {
+        print_topology(&stream_info, 1);
+    }
+}
+
+fn main() -> Result<(), Error> {
+    gst::init()?;
+
+    let args: Vec<_> = env::args().collect();
+    let uri: &str = if args.len() == 2 {
```

Comment:
`>= 2` maybe?

Reply:
Maybe we also just want discoverer all args async instead of just the first?

Reply:
That would be different behaviour to the C tutorial, although arguably a little more interesting. One problem is that I haven't read the concurrency parts of rust which would be required to keep an atomic counter of the remaining uri's to be discovered so that loop.quit() can be properly called once all discoveries have been made. The C API for discovery doesn't appear to give you any calls to check if the loop is idle, so this would need to be added to the Rust code, perhaps complicating it more than necessary to demonstrate usage of the discovery API.

Reply:
Good point. Let's keep it simple then :)